duckduckgo/Android
# Android: New internal Pixel api

Task/Issue URL: https://app.asana.com/0/414730916066338/1198967073193796/f
Tech Design URL: 
CC: 

**Description**:
Currently, our Pixel api relies on a network call to send any pixel. This covers most of our usecases when we need to send a Pixel to our backend.
There are some scenarios where this is not enough:
- Sending pixels when we are about to restart the app (pressing Fire)
- If network fails

This PR introduces a new method in our existing API: enqueuePixel. This method is opt-in, so we can use it in cases where we are afraid about losing pixels and those are important.

`enqueuePixel` will:
- store the pixel in the database, and
- try to send the pixel to remote API
- if succeeds, remote pixel
- if fails, next time another pixel is added or the application comes again to the foreground, try to send it again.

In this iteration, I discarded using a WorkManager to keep it simple and to make it as synchronous as possible (as soon as the pixel is stored in our database, we will try to send it). Making it synchronous will make it easier to test PRs that include any pixel validation step, the reviewer will not need to wait any amount of time to validate if a pixel is sent.

**Steps to test this PR**:

> Fresh install

**Test 1:**
1. Go to any part of the code where we use `pixel.fire`, and replace it for `pixel.enqueueFire`
1. Perform a fresh install
1. Go to the screen where you changed the pixel method call
1. Peform the action required to enqueue the pixel
1. ensure Pixel is sent (see logs `Pixel sent`)
1. ensure in the database, the pixel has been removed

**Test 2:**
1. Turn off data on your emulator or device
1. Go to the screen where you changed the pixel method call
1. Peform the action required to enqueue the pixel
1. ensure Pixel is not sent (we should not see any log saying `Pixel sent`)
1. ensure in the database, all unsent triggered pixels are present
1. Turn on data on your emulator or device
1. Perform any of the following actions: another action that enqueues a new pixel OR send the app to background and bring it back to foreground OR press fire button
1. ensure Pixel is sent (see logs `Pixel sent`)
1. ensure in the database, the pixel has been removed

> Database Migration

**Test 1:**
1. Install develop and run the app
1. Checkout this branch and install it on your device
1. Ensure migration succeeds


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
